ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +175,424 @@ let order_f (_, vm_rec) =
+end
+
+(* For a VM anti-affinity group, the state of a host which determines
+   evacuation planning for anti-affinity VMs in that group:
+   1. vm_cnt: the number of running VMs in that group resident on the host
+   2. h_size: the amount of free memory of the host *)
+module AntiAffinityEvacPlanHostState = struct
+  type t = int * int64
+
+  (* During evacuation planning for anti-affintiy VMs, "vm_cnt" is the first
+     factor considered, "h_size" is the second factor considered:
+     Let's say the next anti-affinity VM to be planned belongs to group A, the
+     host to be selected should be the one which has minimal "vm_cnt" of group
+     A, for hosts with the same "vm_cnt", pick the one with the minimal
+     "h_size" which can hold the VM(h_size >= vm_size). *)
+  let compare (vm_cnt_0, h_size_0) (vm_cnt_1, h_size_1) =
+    match Int.compare vm_cnt_0 vm_cnt_1 with
+    | 0 ->
+        Int64.compare h_size_0 h_size_1
+    | c ->
+        c
+end
+
+(* A Psq of hosts for an anti-affinity group, which is used for evacuation
+   planning for anti-affinity VMs in that group: the minimal host in this Psq
+   is the first one to be considered to plan a VM from that group.
+   When several hosts share the minimal "AntiAffinityEvacPlanHostState",
+   the minimal host is the host with the smallest ref. *)
+module AntiAffinityEvacPlanHostPsq =
+  Psq.Make (HostKey) (AntiAffinityEvacPlanHostState)
+module Q = AntiAffinityEvacPlanHostPsq
+
+(* The spread evenly plan pool state determines the spread evenly evacuation
+   planning for anti-affinity VMs.
+   It's a VMGrpMap which contains a Psq Q for each group, and each Psq contains
+   all the available hosts in the pool.
+   Let's say the anti-affinity VM to be planned belongs to anti-affinity group
+   A. To get a spread evenly evacuation plan, the most suitable host to plan
+   the VM would be the host which has the minimal number of running VMs from
+   group A resident on it, for the hosts with the same number of running VMs
+   from group A, the one with the minimal free memory will be checked first,
+   which is just the minimal host returned from "Q.min" on the Psq of group A.
+*)
+let init_spread_evenly_plan_pool_state ~__context hosts =
+  let gen_psq host_vm_cnt =
+    List.fold_left
+      (fun q (h, h_size) ->
+        let vm_cnt =
+          Xapi_vm_helpers.HostMap.find_opt h host_vm_cnt
+          |> Option.value ~default:0
         in
-        (vm, total_memory_of_vm ~__context policy snapshot)
+        Q.add h (vm_cnt, h_size) q
       )
-      vms_and_snapshots
+      Q.empty hosts
   in
+  Db.VM_group.get_all ~__context
+  |> List.filter (fun group ->
+         Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
+     )
+  |> List.fold_left
+       (fun grp_psq grp ->
+         let host_vm_cnt =
+           Xapi_vm_helpers.host_to_vm_count_map ~__context grp
+         in
+         VMGrpMap.add grp (gen_psq host_vm_cnt) grp_psq
+       )
+       VMGrpMap.empty
```

Comment:
Without the change on the "retrieve groups from VMs to be migrated instead of all groups in DB":
```
let init_spread_evenly_plan_pool_state ~__context hosts =
  let module H = Xapi_vm_helpers in
  let gen_psq grp =
    let m = H.host_to_vm_count_map ~__context grp in
    hosts
    |> List.map
        (fun (h, h_size) ->
          let vm_cnt = H.HostMap.find_opt h m |> Option.value ~default:0 in
          (h, (vm_cnt, h_size))
        )
    |> Q.of_list
  in
  Db.VM_group.get_all ~__context
  |> List.filter (fun group ->
         Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
     )
  |> List.fold_left
      (fun add grp -> VMGrpMap.add grp (gen_psq grp) acc)
      VMGrpMap.empty
```